File: pkg/memory/constraint.go
package memory

import (
	"fmt"
	"sync"
	"sync/atomic"
)

// Constraint References - Assertion-based safety for complex patterns
//
// For graphs, observers, callbacks, and other complex ownership patterns.
//
// Key concept:
// - Each object has ONE owner (responsible for freeing)
// - Multiple non-owning "constraint" references can exist
// - On free: ASSERT that constraint count is zero
// - If assertion fails: dangling references would have been created
//
// This is primarily a DEBUG/DEVELOPMENT tool - catches errors at free time
// rather than at dereference time.

// ConstraintObj is an object with constraint reference tracking
type ConstraintObj struct {
	Data            interface{}
	Owner           string // Description of owner for debugging
	ConstraintCount int32  // Number of active constraint references (atomic)
	Freed           bool   // Whether the object has been freed

	// Debug info - only used when debug tracking is enabled
	mu                sync.Mutex
	ConstraintSources map[uint64]string // refID -> source (only in debug mode)
	nextRefID         uint64
	debugMode         bool
}

// ConstraintRef is a non-owning reference that constrains object lifetime
type ConstraintRef struct {
	Target   *ConstraintObj
	Source   string // Where this ref was created
	refID    uint64 // Unique ID for O(1) removal
	released int32  // Atomic flag
}

// ConstraintContext manages constraint objects
type ConstraintContext struct {
	Objects       []*ConstraintObj
	AssertOnError bool     // If true, panic on constraint violation
	Violations    []string // Recorded violations if not asserting
	DebugMode     bool     // Enable source tracking (slower but more info)
	mu            sync.Mutex
}

// NewConstraintContext creates a new constraint context
func NewConstraintContext(assertOnError bool) *ConstraintContext {
	return &ConstraintContext{
		Objects:       make([]*ConstraintObj, 0),
		AssertOnError: assertOnError,
		Violations:    make([]string, 0),
		DebugMode:     false, // Off by default for performance
	}
}

// NewConstraintContextDebug creates a context with debug tracking enabled
func NewConstraintContextDebug(assertOnError bool) *ConstraintContext {
	return &ConstraintContext{
		Objects:       make([]*ConstraintObj, 0),
		AssertOnError: assertOnError,
		Violations:    make([]string, 0),
		DebugMode:     true,
	}
}

// Alloc allocates a new constraint object with an owner
func (ctx *ConstraintContext) Alloc(data interface{}, owner string) *ConstraintObj {
	obj := &ConstraintObj{
		Data:      data,
		Owner:     owner,
		debugMode: ctx.DebugMode,
	}
	if ctx.DebugMode {
		obj.ConstraintSources = make(map[uint64]string)
	}
	ctx.mu.Lock()
	ctx.Objects = append(ctx.Objects, obj)
	ctx.mu.Unlock()
	return obj
}

// AddConstraint creates a constraint reference to an object
// The constraint must be released before the object can be freed
func (obj *ConstraintObj) AddConstraint(source string) (*ConstraintRef, error) {
	obj.mu.Lock()
	if obj.Freed {
		obj.mu.Unlock()
		return nil, fmt.Errorf("cannot add constraint to freed object")
	}

	// Increment constraint count
	atomic.AddInt32(&obj.ConstraintCount, 1)

	ref := &ConstraintRef{
		Target: obj,
		Source: source,
	}

	// Only track sources in debug mode
	if obj.debugMode {
		obj.nextRefID++
		ref.refID = obj.nextRefID
		obj.ConstraintSources[ref.refID] = source
	}
	obj.mu.Unlock()

	return ref, nil
}

// Release releases a constraint reference
func (ref *ConstraintRef) Release() error {
	// Atomic check-and-set for released flag
	if !atomic.CompareAndSwapInt32(&ref.released, 0, 1) {
		return fmt.Errorf("constraint already released [%s]", ref.Source)
	}
	if ref.Target == nil {
		return fmt.Errorf("null constraint reference")
	}

	// Atomically decrement constraint count
	newCount := atomic.AddInt32(&ref.Target.ConstraintCount, -1)
	if newCount < 0 {
		// Restore and return error
		atomic.AddInt32(&ref.Target.ConstraintCount, 1)
		atomic.StoreInt32(&ref.released, 0)
		return fmt.Errorf("constraint count underflow")
	}

	// Remove from sources map only in debug mode
	if ref.Target.debugMode {
		ref.Target.mu.Lock()
		delete(ref.Target.ConstraintSources, ref.refID)
		ref.Target.mu.Unlock()
	}

	return nil
}

// Free attempts to free a constraint object
// Returns error if constraints still exist
func (ctx *ConstraintContext) Free(obj *ConstraintObj) error {
	obj.mu.Lock()
	defer obj.mu.Unlock()

	if obj.Freed {
		return fmt.Errorf("double free detected [owner: %s]", obj.Owner)
	}

	count := atomic.LoadInt32(&obj.ConstraintCount)
	if count > 0 {
		var sources []string
		if obj.debugMode {
			sources = make([]string, 0, len(obj.ConstraintSources))
			for _, src := range obj.ConstraintSources {
				sources = append(sources, src)
			}
		}

		violation := fmt.Sprintf(
			"constraint violation: cannot free object [owner: %s] with %d active constraints",
			obj.Owner, count,
		)
		if len(sources) > 0 {
			violation += fmt.Sprintf(" from: %v", sources)
		}

		ctx.mu.Lock()
		ctx.Violations = append(ctx.Violations, violation)
		ctx.mu.Unlock()

		if ctx.AssertOnError {
			panic(violation)
		}
		return fmt.Errorf(violation)
	}

	obj.Freed = true
	obj.Data = nil
	return nil
}

// MustFree frees the object, panicking if constraints exist
func (ctx *ConstraintContext) MustFree(obj *ConstraintObj) {
	err := ctx.Free(obj)
	if err != nil {
		panic(err)
	}
}

// Deref safely dereferences a constraint reference
func (ref *ConstraintRef) Deref() (interface{}, error) {
	if ref.Target == nil {
		return nil, fmt.Errorf("null constraint reference")
	}
	if atomic.LoadInt32(&ref.released) == 1 {
		return nil, fmt.Errorf("dereferencing released constraint [%s]", ref.Source)
	}
	ref.Target.mu.Lock()
	if ref.Target.Freed {
		ref.Target.mu.Unlock()
		return nil, fmt.Errorf("use-after-free: object was freed [owner: %s, ref: %s]",
			ref.Target.Owner, ref.Source)
	}
	data := ref.Target.Data
	ref.Target.mu.Unlock()
	return data, nil
}

// IsValid checks if a constraint reference is valid
func (ref *ConstraintRef) IsValid() bool {
	if ref.Target == nil {
		return false
	}
	if atomic.LoadInt32(&ref.released) == 1 {
		return false
	}
	ref.Target.mu.Lock()
	freed := ref.Target.Freed
	ref.Target.mu.Unlock()
	return !freed
}

// GetViolations returns all recorded violations
func (ctx *ConstraintContext) GetViolations() []string {
	ctx.mu.Lock()
	defer ctx.mu.Unlock()
	result := make([]string, len(ctx.Violations))
	copy(result, ctx.Violations)
	return result
}

// HasViolations checks if any violations occurred
func (ctx *ConstraintContext) HasViolations() bool {
	ctx.mu.Lock()
	defer ctx.mu.Unlock()
	return len(ctx.Violations) > 0
}

// ClearViolations clears the violation log
func (ctx *ConstraintContext) ClearViolations() {
	ctx.mu.Lock()
	defer ctx.mu.Unlock()
	ctx.Violations = ctx.Violations[:0]
}

// ConstraintStats provides statistics
type ConstraintStats struct {
	TotalObjects     int
	ActiveObjects    int
	TotalConstraints int
	ViolationCount   int
}

// GetStats returns current statistics
func (ctx *ConstraintContext) GetStats() ConstraintStats {
	ctx.mu.Lock()
	defer ctx.mu.Unlock()

	active := 0
	totalConstraints := 0
	for _, obj := range ctx.Objects {
		obj.mu.Lock()
		if !obj.Freed {
			active++
			totalConstraints += int(atomic.LoadInt32(&obj.ConstraintCount))
		}
		obj.mu.Unlock()
	}

	return ConstraintStats{
		TotalObjects:     len(ctx.Objects),
		ActiveObjects:    active,
		TotalConstraints: totalConstraints,
		ViolationCount:   len(ctx.Violations),
	}
}

File: runtime/src/memory/constraint.c
/*
 * Constraint References implementation
 * Assertion-based safety for complex patterns
 */

#include "constraint.h"
#include <stdlib.h>
#include <stdio.h>
#include <string.h>

/* Helper: grow array */
static void* grow_array(void* arr, int* capacity, size_t elem_size) {
    int new_cap = (*capacity == 0) ? 8 : (*capacity * 2);
    void* new_arr = realloc(arr, new_cap * elem_size);
    if (new_arr) {
        *capacity = new_cap;
    }
    return new_arr;
}

/* Helper: add violation to context */
static void add_violation(ConstraintContext* ctx, const char* message) {
    if (!ctx) return;

    /* Duplicate the message */
    char* msg_copy = strdup(message);
    if (!msg_copy) return;

    if (ctx->violation_count >= ctx->violation_capacity) {
        char** new_violations = grow_array(
            ctx->violations,
            &ctx->violation_capacity,
            sizeof(char*)
        );
        if (!new_violations) {
            free(msg_copy);
            return;
        }
        ctx->violations = new_violations;
    }
    ctx->violations[ctx->violation_count++] = msg_copy;
}

/* Create new context */
ConstraintContext* constraint_context_new(bool assert_on_error) {
    ConstraintContext* ctx = calloc(1, sizeof(ConstraintContext));
    if (!ctx) return NULL;

    ctx->objects = NULL;
    ctx->object_count = 0;
    ctx->object_capacity = 0;
    ctx->assert_on_error = assert_on_error;
    ctx->violations = NULL;
    ctx->violation_count = 0;
    ctx->violation_capacity = 0;

    return ctx;
}

/* Free context */
void constraint_context_free(ConstraintContext* ctx) {
    if (!ctx) return;

    /* Free violation strings */
    for (int i = 0; i < ctx->violation_count; i++) {
        free(ctx->violations[i]);
    }
    free(ctx->violations);
    free(ctx->objects);
    free(ctx);
}

/* Allocate new constraint object */
ConstraintObj* constraint_alloc(ConstraintContext* ctx, void* data,
                                void (*destructor)(void*), const char* owner) {
    ConstraintObj* obj = calloc(1, sizeof(ConstraintObj));
    if (!obj) return NULL;

    obj->data = data;
    obj->destructor = destructor;
    obj->owner = owner;
    obj->constraint_count = 0;
    obj->freed = false;
    obj->source_count = 0;

    /* Add to context if provided */
    if (ctx) {
        if (ctx->object_count >= ctx->object_capacity) {
            ConstraintObj** new_objs = grow_array(
                ctx->objects,
                &ctx->object_capacity,
                sizeof(ConstraintObj*)
            );
            if (!new_objs) {
                free(obj);
                return NULL;
            }
            ctx->objects = new_objs;
        }
        ctx->objects[ctx->object_count++] = obj;
    }

    return obj;
}

/* Add constraint reference to object */
ConstraintRef* constraint_add(ConstraintObj* obj, const char* source) {
    if (!obj || obj->freed) return NULL;

    ConstraintRef* ref = calloc(1, sizeof(ConstraintRef));
    if (!ref) return NULL;

    ref->target = obj;
    ref->source = source;
    ref->released = false;

    obj->constraint_count++;

    /* Track source for debugging */
    if (obj->source_count < MAX_CONSTRAINT_SOURCES) {
        obj->constraint_sources[obj->source_count++] = source;
    }

    return ref;
}

/* Release constraint reference */
ConstraintError constraint_release(ConstraintRef* ref) {
    if (!ref) return CONSTRAINT_ERR_NULL;
    if (ref->released) return CONSTRAINT_ERR_DOUBLE_RELEASE;
    if (!ref->target) return CONSTRAINT_ERR_NULL;

    if (ref->target->constraint_count <= 0) {
        return CONSTRAINT_ERR_VIOLATION;
    }

    ref->target->constraint_count--;
    ref->released = true;

    /* Remove from sources list */
    for (int i = 0; i < ref->target->source_count; i++) {
        if (ref->target->constraint_sources[i] == ref->source) {
            /* Shift remaining sources */
            for (int j = i; j < ref->target->source_count - 1; j++) {
                ref->target->constraint_sources[j] = ref->target->constraint_sources[j + 1];
            }
            ref->target->source_count--;
            break;
        }
    }

    return CONSTRAINT_OK;
}

/* Free constraint reference struct */
void constraint_ref_free(ConstraintRef* ref) {
    free(ref);
}

/* Free constraint object */
ConstraintError constraint_free(ConstraintContext* ctx, ConstraintObj* obj) {
    if (!obj) return CONSTRAINT_ERR_NULL;
    if (obj->freed) return CONSTRAINT_ERR_DOUBLE_FREE;

    if (obj->constraint_count > 0) {
        /* Build violation message */
        char msg[512];
        snprintf(msg, sizeof(msg),
                 "constraint violation: cannot free object [owner: %s] with %d active constraints",
                 obj->owner ? obj->owner : "unknown",
                 obj->constraint_count);

        if (ctx) {
            add_violation(ctx, msg);

            if (ctx->assert_on_error) {
                fprintf(stderr, "%s\n", msg);
                fprintf(stderr, "  Constraint sources:\n");
                for (int i = 0; i < obj->source_count; i++) {
                    fprintf(stderr, "    - %s\n", obj->constraint_sources[i]);
                }
                abort();
            }
        }

        return CONSTRAINT_ERR_VIOLATION;
    }

    obj->freed = true;
    if (obj->destructor && obj->data) {
        obj->destructor(obj->data);
    }
    obj->data = NULL;

    return CONSTRAINT_OK;
}

/* Dereference safely */
void* constraint_deref(ConstraintRef* ref, ConstraintError* err) {
    if (!ref || !ref->target) {
        if (err) *err = CONSTRAINT_ERR_NULL;
        return NULL;
    }
    if (ref->released) {
        if (err) *err = CONSTRAINT_ERR_RELEASED;
        return NULL;
    }
    if (ref->target->freed) {
        if (err) *err = CONSTRAINT_ERR_FREED;
        return NULL;
    }

    if (err) *err = CONSTRAINT_OK;
    return ref->target->data;
}

/* Check if reference is valid */
bool constraint_ref_is_valid(ConstraintRef* ref) {
    return ref && ref->target && !ref->released && !ref->target->freed;
}

/* Check if context has violations */
bool constraint_has_violations(ConstraintContext* ctx) {
    return ctx && ctx->violation_count > 0;
}

/* Get violation count */
int constraint_get_violation_count(ConstraintContext* ctx) {
    return ctx ? ctx->violation_count : 0;
}

/* Get violation by index */
const char* constraint_get_violation(ConstraintContext* ctx, int index) {
    if (!ctx || index < 0 || index >= ctx->violation_count) return NULL;
    return ctx->violations[index];
}

/* Clear violations */
void constraint_clear_violations(ConstraintContext* ctx) {
    if (!ctx) return;
    for (int i = 0; i < ctx->violation_count; i++) {
        free(ctx->violations[i]);
    }
    ctx->violation_count = 0;
}

/* Get statistics */
ConstraintStats constraint_get_stats(ConstraintContext* ctx) {
    ConstraintStats stats = {0, 0, 0, 0};
    if (!ctx) return stats;

    stats.total_objects = ctx->object_count;
    stats.violation_count = ctx->violation_count;

    for (int i = 0; i < ctx->object_count; i++) {
        ConstraintObj* obj = ctx->objects[i];
        if (obj && !obj->freed) {
            stats.active_objects++;
            stats.total_constraints += obj->constraint_count;
        }
    }

    return stats;
}
